### PR TITLE
Adding dynamic client & deploymentconfig object support

### PR DESCRIPTION
**Pull request by rajdas98**
Signed-off-by: Raj <raj.das@mayadata.io>
Issue- https://github.com/litmuschaos/litmus/issues/1406

- Adding Openshift's Deploymentconfig annotation check support.
- Adding dynamic client for using openshift & k8s resources

### CODE REVIEW

**Comment by rajdas98** [comment]
```
{"level":"info","ts":1588442791.1885405,"logger":"cmd","msg":"Starting the Chaos-Operator..."}
{"level":"info","ts":1588442791.188674,"logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}
{"level":"info","ts":1588442791.288929,"logger":"controller-runtime.controller","msg":"Starting Controller","controller":"chaosengine-controller"}
{"level":"info","ts":1588442791.3893805,"logger":"controller-runtime.controller","msg":"Starting workers","controller":"chaosengine-controller","worker count":1}
{"level":"info","ts":1588442806.8061612,"logger":"controller_chaosengine","msg":"Reconciling ChaosEngine","Request.Namespace":"litmus","Request.Name":"nginx-chaos"}
Hello dc
{"level":"info","ts":1588442806.81568,"logger":"controller_chaosengine","msg":"App key derived from chaosengine is ","appLabelKey":"run"}
{"level":"info","ts":1588442806.8157,"logger":"controller_chaosengine","msg":"App Label derived from Chaosengine is ","appLabelValue":"nginx"}
{"level":"info","ts":1588442806.8157034,"logger":"controller_chaosengine","msg":"App NS derived from Chaosengine is ","appNamespace":"default"}
{"level":"info","ts":1588442806.815715,"logger":"controller_chaosengine","msg":"Exp list derived from chaosengine is ","appExpirements":["pod-delete"]}
{"level":"info","ts":1588442806.815722,"logger":"controller_chaosengine","msg":"Monitoring Status derived from chaosengine is","monitoringStatus":false}
{"level":"info","ts":1588442806.8157256,"logger":"controller_chaosengine","msg":"Runner image derived from chaosengine is","runnerImage":"litmuschaos/chaos-runner:1.3.0"}
{"level":"info","ts":1588442806.8157284,"logger":"controller_chaosengine","msg":"Annotation check is ","annotationCheck":"true"}
{"level":"info","ts":1588442806.8300862,"logger":"controller_chaosengine","msg":"Deployment chaos candidate:","appName: ":"nginx"," appUUID: ":"f487c9d1-8c7b-11ea-99b3-080027b410b8"}
{"level":"info","ts":1588442806.8301163,"logger":"controller_chaosengine","msg":"chaos candidate of","kind:":"deploymentconfig","appName: ":"nginx","appUUID: ":"f487c9d1-8c7b-11ea-99b3-080027b410b8"}
{"level":"info","ts":1588442806.830149,"logger":"controller_chaosengine","msg":"Creating a new engineRunner Pod","Request.Namespace":"litmus","Request.Name":"nginx-chaos","Pod.Namespace":"litmus","Pod.Name":"nginx-chaos-runner"}
ChaosEnabled 1
{"level":"info","ts":1588442806.8475034,"logger":"controller_chaosengine","msg":"engineRunner Pod created successfully","Request.Namespace":"litmus","Request.Name":"nginx-chaos"}
{"level":"info","ts":1588442806.847521,"logger":"controller_chaosengine","msg":"Skip reconcile: engineRunner Pod already exists","Request.Namespace":"litmus","Request.Name":"nginx-chaos","Pod.Namespace":"","Pod.Name":""}
{"level":"info","ts":1588442806.8475668,"logger":"controller_chaosengine","msg":"Reconciling ChaosEngine","Request.Namespace":"litmus","Request.Name":"nginx-chaos"}
{"level":"info","ts":1588442806.8475826,"logger":"controller_chaosengine","msg":"App key derived from chaosengine is ","appLabelKey":"run"}
{"level":"info","ts":1588442806.8475857,"logger":"controller_chaosengine","msg":"App Label derived from Chaosengine is ","appLabelValue":"nginx"}
{"level":"info","ts":1588442806.8475888,"logger":"controller_chaosengine","msg":"App NS derived from Chaosengine is ","appNamespace":"default"}
{"level":"info","ts":1588442806.8475916,"logger":"controller_chaosengine","msg":"Exp list derived from chaosengine is ","appExpirements":["pod-delete"]}
{"level":"info","ts":1588442806.8475964,"logger":"controller_chaosengine","msg":"Monitoring Status derived from chaosengine is","monitoringStatus":false}
{"level":"info","ts":1588442806.8475995,"logger":"controller_chaosengine","msg":"Runner image derived from chaosengine is","runnerImage":"litmuschaos/chaos-runner:1.3.0"}
{"level":"info","ts":1588442806.8476024,"logger":"controller_chaosengine","msg":"Annotation check is ","annotationCheck":"true"}
Hello dc
{"level":"info","ts":1588442806.8733604,"logger":"controller_chaosengine","msg":"Deployment chaos candidate:","appName: ":"nginx"," appUUID: ":"f487c9d1-8c7b-11ea-99b3-080027b410b8"}
{"level":"info","ts":1588442806.8733869,"logger":"controller_chaosengine","msg":"chaos candidate of","kind:":"deploymentconfig","appName: ":"nginx","appUUID: ":"f487c9d1-8c7b-11ea-99b3-080027b410b8"}
```

File: deploy/01-chaos-operator.yaml [hunk 1]
```diff
@@ -19,9 +19,9 @@ metadata:
   labels:
     name: litmus
 rules:
-- apiGroups: ["","apps","batch","litmuschaos.io"]
-  resources: ["pods","jobs","deployments","daemonsets","statefulsets","events","configmaps","services","chaosengines","chaosexperiments","chaosresults"]
-  verbs: ["get","create","update","patch","delete","list","watch","deletecollection"]
+  - apiGroups: ["","apps","batch","litmuschaos.io","openshift.io"]
+    resources: ["pods","jobs","deployments","daemonsets","replicaset","deploymentconfig","statefulsets","events","configmaps","services","chaosengines","chaosexperiments","chaosresults"]
```
Comment (rahulchheda) [inline deploy/01-chaos-operator.yaml:23]:
```suggestion
    resources: ["pods","jobs","deployments","daemonsets","replicasets","deploymentconfigs","statefulsets","events","configmaps","services","chaosengines","chaosexperiments","chaosresults"]
```

File: deploy/01-chaos-operator.yaml [hunk 2]
```diff
@@ -316,9 +316,9 @@ spec:
           type: object
   version: v1alpha1
   versions:
-  - name: v1alpha1
-    served: true
-    storage: true
+    - name: v1alpha1
```
Comment (rahulchheda) [inline deploy/01-chaos-operator.yaml:319]: Are the spaces added via linting tool or self added?

File: pkg/controller/resource/deploymentconfig.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	chaosTypes "github.com/litmuschaos/chaos-operator/pkg/controller/types"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:11]:
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/client-go/dynamic"
	
	chaosTypes "github.com/litmuschaos/chaos-operator/pkg/controller/types"
	
```

File: pkg/controller/resource/deploymentconfig.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	engine, chaosEnabledDeployment, err := checkForChaosEnabledDeploymentConfig(targetAppList, engine)
+	if err != nil {
+		return engine, err
+	}
+
+	if chaosEnabledDeployment == 0 {
+		return engine, errors.New("no chaos-candidate found")
+	}
+	chaosTypes.Log.Info("Deployment chaos candidate:", "appName: ", engine.AppName, " appUUID: ", engine.AppUUID)
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:37]:
```suggestion
	chaosTypes.Log.Info("Deployment Config chaos candidate:", "appName: ", engine.AppName, " appUUID: ", engine.AppUUID)
```

File: pkg/controller/resource/deploymentconfig.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+	return engine, nil
+}
+
+func getDeploymentConfigLists(clientSet dynamic.Interface, engine *chaosTypes.EngineInfo) (*unstructured.UnstructuredList, error) {
+
+	dyn := clientSet.Resource(dcGVR)
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:44]: Please add a suitable name to this variable, which is self explanatory

File: pkg/controller/resource/deploymentconfig.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	targetAppList, err := dyn.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:47]: Please add suitable error message like we have for other resources

File: pkg/controller/resource/deploymentconfig.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:49]: Add check for empty items in this function

File: pkg/controller/resource/deploymentconfig.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+func checkForChaosEnabledDeploymentConfig(targetAppList *unstructured.UnstructuredList, engine *chaosTypes.EngineInfo) (*chaosTypes.EngineInfo, int, error) {
+
+	chaosEnabledDeployment := 0
+	for _, deploymentconfig := range targetAppList.Items {
+		engine.AppName = deploymentconfig.GetName()
+		engine.AppUUID = deploymentconfig.GetUID()
+		annotationValue := deploymentconfig.GetAnnotations()[ChaosAnnotationKey]
+		chaosEnabledDeployment = CountTotalChaosEnabled(annotationValue, chaosEnabledDeployment)
+		if chaosEnabledDeployment > 1 {
+			return engine, chaosEnabledDeployment, errors.New("too many deployments with specified label are annotated for chaos, either provide unique labels or annotate only desired app for chaos")
```
Comment (rahulchheda) [inline pkg/controller/resource/deploymentconfig.go:63]:
```suggestion
			return engine, chaosEnabledDeployment, errors.New("too many deployment configs with specified label are annotated for chaos, either provide unique labels or annotate only desired app for chaos")
```

File: deploy/01-chaos-operator.yaml (hunk 2 above)
Comment (rajdas98) [reply deploy/01-chaos-operator.yaml:319]: Lint

**Comment by rajdas98** [comment]
Closing this PR to re-run the circleci. Raised https://github.com/litmuschaos/chaos-operator/pull/199

**Pull request closed without merging**